src/coprocessor/dag/expr/arithmetic.rs

```
@@ -0,0 +1,390 @@
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() - r.as_ref()).into();
+            result.map(Cow::Owned)
+        })
+    }
+
+    pub fn minus_int(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| l.checked_sub(r).ok_or(Error::Overflow))
+    }
+
+    pub fn minus_uint(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<i64>> {
+        let lhs = try!(self.children[0].eval_int(ctx, row));
+        let rhs = try!(self.children[1].eval_int(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = (l as u64).checked_sub(r as u64).ok_or(Error::Overflow);
+            res.map(|u| u as i64)
+        })
+    }
+
+    pub fn multiply_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try!(self.children[0].eval_real(ctx, row));
+        let rhs = try!(self.children[1].eval_real(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let res = l * r;
+            if !res.is_finite() {
+                return Err(Error::Overflow);
+            }
+            Ok(res)
+        })
+    }
+
+    pub fn multiply_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try!(self.children[1].eval_decimal(ctx, row));
+        do_arithmetic(lhs, rhs, |l, r| {
+            let result: Result<Decimal> = (l.as_ref() * r.as_ref()).into();
```

I believe the error is dropped in tidb's implementation.